### PR TITLE
Get Profile By Id

### PR DESCRIPTION
**Pull request by Mari-Hayashi**
Implemented two backend APIs
    -   api/student/get-profile
    -   api/facultyMember/get-profile
The APIs returns the user profile based on the id.

### CODE REVIEW

File: src/server/README.md [hunk 1]
```diff
@@ -386,6 +386,52 @@ Faculty Member API
...
+            -   success:
+                Status code: 200
+                ```
+                {
+                    facultyMember: {
+                        id: number,
+                        user:{
+                            id: number,
+                            email: : string,
+                            password: : string, // Encoded
```
Comment (johan1505) [inline src/server/README.md]: lol I don't think we should give back the password. I know it's encoded but it's not a good practice

File: src/server/README.md [hunk 2]
```diff
@@ -434,7 +480,57 @@ Student API
...
+                ```
+                {
+                    student: {
+                        id: number,
+                        sid: number,
+                        classStanding: 'freshman' | 'sophomore' | 'junior' | 'senior',
+                        user: {
+                            id: number,
+                            email: : string,
+                            password: : string, // Encoded
```
Comment (johan1505) [inline src/server/README.md]: Same here

File: src/server/README.md [hunk 3]
```diff
@@ -434,7 +480,57 @@ Student API
...
+                        department: {
+                            id: number,
+                            name: string,
+                            college: {
+                                id: number,
+                                name: string
+                            }
+                        },
+                        courses: {
+                                "id": 5,
```
Comment (johan1505) [inline src/server/README.md]: add integer type and remove 5

File: src/server/README.md [hunk 4]
```diff
@@ -434,7 +480,57 @@ Student API
...
+                                "title": "CS200"
```
Comment (johan1505) [inline src/server/README.md]: add string type and remove "CS200"

File: src/server/src/routes/facultyMember.ts [hunk 5]
```diff
@@ -74,6 +74,30 @@ router.post('/update-profile', async (req: facultyMemberRequest, res: Response)
     }
 });
 
+/******************************************************************************
+ *          GET Request - Read - "GET /api/facultyMember/get-profile"
+ ******************************************************************************/
+
+router.get('/get-profile', async (req: Request, res: Response) => {
+    const { id } = req.body;
```
Comment (johan1505) [inline src/server/src/routes/facultyMember.ts]: The facultyMemberId can be received from the user object found in the request (as long as the request comes with a cookie, i.e user is authenticated). You can find an example of how to do this in the PR I assigned you, in the file routes/job.ts line 32 - 35.

File: src/server/src/routes/student.ts [hunk 6]
```diff
@@ -73,6 +74,30 @@ router.post('/update-profile', async (req: studentRequest, res: Response) => {
     }
 });
 
+/******************************************************************************
+ *          GET Request - Read - "GET /api/student/get-profile"
+ ******************************************************************************/
+
+router.get('/get-profile', async (req: Request, res: Response) => {
+    const { id } = req.body;
```
Comment (johan1505) [inline src/server/src/routes/student.ts:82]: Same here studentId be received from the user object found in the request (as long as the request comes with a cookie, i.e user is authenticated). Note you would have to use the JWTStudent interface (declared in entities/user.ts). LMK if this makes sense.

**Review by johan1505 — CHANGES_REQUESTED** [review]
Left comments.

File: src/server/src/routes/facultyMember.ts (hunk 5 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/facultyMember.ts]: I thought we also want to use this API to let the users look at other people's profiles. (Like when a student wants to look at a faculty member's profile.), so I don't think we always want to get id from the login information. We can create two separate APIs, one that is already implemented and another one that returns the profile of the logged-in user. What do you think?

Comment (johan1505) [reply src/server/src/routes/facultyMember.ts]: ah I see, then this is fine. But the callers of these APIs should be authenticated. We just don't care about their role (student or faculty member). Also, make sure you don't return sensitive information in the profile back in the response.

File: src/server/README.md (hunk 1 above)
Comment (Mari-Hayashi) [reply src/server/README.md]: Done.

File: src/server/README.md (hunk 2 above)
Comment (Mari-Hayashi) [reply src/server/README.md]: Done.

File: src/server/README.md (hunk 3 above)
Comment (Mari-Hayashi) [reply src/server/README.md]: Done.

File: src/server/README.md (hunk 4 above)
Comment (Mari-Hayashi) [reply src/server/README.md]: Done.

File: src/server/src/routes/facultyMember.ts (hunk 5 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/facultyMember.ts]: Done.

File: src/server/src/routes/facultyMember.ts [hunk 7]
```diff
@@ -82,24 +84,34 @@ router.post(
...
-        logger.err(error);
-        return res
-            .status(INTERNAL_SERVER_ERROR)
-            .json(errors.internalServerError)
-            .end();
-    }
+
+        const { id } = req.body;
+        try {
+            if (!id) {
```
Comment (johan1505) [inline src/server/src/routes/facultyMember.ts]: if statements should be outside of try-catch block

File: src/server/src/routes/facultyMember.ts [hunk 8]
```diff
@@ -82,24 +84,34 @@ router.post(
...
-        return res.status(OK).json({ facultyMember }).end();
-    } catch (error) {
-        logger.err(error);
-        return res
-            .status(INTERNAL_SERVER_ERROR)
-            .json(errors.internalServerError)
-            .end();
-    }
+
+        const { id } = req.body;
```
Comment (johan1505) [inline src/server/src/routes/facultyMember.ts]: id doesn't have a type

File: src/server/src/routes/student.ts [hunk 9]
```diff
@@ -71,24 +73,32 @@ router.post('/update-profile', async (req: studentRequest, res: Response) => {
...
+router.get('/get-profile',
+    passport.authenticate('jwt', { session: false }),
+    async (req: Request, res: Response) => {
+        const { userId } = req.user as JWTUser;
+        if (!userId) {
+            return res
+                .status(UNAUTHORIZED)
+                .json({ error: 'You should log in to your account to view the page.' });
         }
+        const { id } = req.body;
```
Comment (johan1505) [inline src/server/src/routes/student.ts]: id doesn't have a type

File: src/server/src/routes/student.ts [hunk 10]
```diff
@@ -71,24 +73,32 @@ router.post('/update-profile', async (req: studentRequest, res: Response) => {
...
+        try {
+            if (!id) {
```
Comment (johan1505) [inline src/server/src/routes/student.ts]: if should be outside of try-catch block

File: src/server/src/routes/facultyMember.ts [hunk 11]
```diff
@@ -82,24 +84,34 @@ router.post(
...
-    const { id } = req.body;
-    try {
-        if (!id) {
-            return res.status(BAD_REQUEST).json({
-                error: errors.paramMissingError,
-            });
+router.get('/get-profile',
+    passport.authenticate('jwt', { session: false }),
+    async (req: Request, res: Response) => {
+        const { userId } = req.user as JWTUser;
```
Comment (johan1505) [inline src/server/src/routes/facultyMember.ts]: No need to check if there is an userId. As long as you put **passport.authenticate('jwt', { session: false })** in the middleware section of the request (like you did), the route will be protected and internally return an UNAUTHORIZED status if the user is not logged in.

File: src/server/src/routes/student.ts [hunk 12]
```diff
@@ -71,24 +73,32 @@ router.post('/update-profile', async (req: studentRequest, res: Response) => {
...
-    const { id } = req.body;
-    try {
-        if (!id) {
-            return res.status(BAD_REQUEST).json({
-                error: errors.paramMissingError,
-            });
+router.get('/get-profile',
+    passport.authenticate('jwt', { session: false }),
+    async (req: Request, res: Response) => {
+        const { userId } = req.user as JWTUser;
```
Comment (johan1505) [inline src/server/src/routes/student.ts]: same here. No need to check for userId

**Review by johan1505 — CHANGES_REQUESTED** [review]
Left more comments

File: src/server/src/routes/facultyMember.ts (hunk 7 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/facultyMember.ts]: Done.

File: src/server/src/routes/facultyMember.ts (hunk 8 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/facultyMember.ts]: Done.

File: src/server/src/routes/student.ts (hunk 9 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/student.ts]: Done.

File: src/server/src/routes/student.ts (hunk 10 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/student.ts]: Done.

File: src/server/src/routes/facultyMember.ts (hunk 11 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/facultyMember.ts]: Done.

File: src/server/src/routes/student.ts (hunk 12 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/student.ts]: Done.

File: src/server/src/routes/student.ts [hunk 13]
```diff
@@ -89,21 +83,23 @@ router.post('/update-profile', async (req: studentRequest, res: Response) => {
...
+ *          GET Request - Read - "GET /api/student/get-profile/:studentId"
  ******************************************************************************/
 
-router.get('/get-profile',
+router.get('/get-profile/:studentId',
     passport.authenticate('jwt', { session: false }),
-    async (req: studentProfileRequest, res: Response) => {
-        const { studentId } = req.body;
+    async (req: Request, res: Response) => {
+        console.log("?");
```
Comment (johan1505) [inline src/server/src/routes/student.ts]: remove console.log

File: src/server/src/routes/student.ts [hunk 14]
```diff
@@ -89,21 +83,23 @@ router.post('/update-profile', async (req: studentRequest, res: Response) => {
...
+        const { studentId } = req.params;
 
+        console.log("?dd");
```
Comment (johan1505) [inline src/server/src/routes/student.ts]: remove console.log

**Review by johan1505 — CHANGES_REQUESTED** [review]
Minor changes

File: src/server/src/routes/student.ts (hunk 13 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/student.ts]: Done.

File: src/server/src/routes/student.ts (hunk 14 above)
Comment (Mari-Hayashi) [reply src/server/src/routes/student.ts]: Done.

**Pull request merged**